src/cloudflare/internal/pipeline-transform.ts

```diff
@@ -0,0 +1,132 @@
+// Copyright (c) 2024 Cloudflare, Inc.
+// Licensed under the Apache 2.0 license found in the LICENSE file or at:
+//     https://opensource.org/licenses/Apache-2.0
+
+import entrypoints from 'cloudflare-internal:workers'
+
+async function* readLines(stream: ReadableStream<string>): AsyncGenerator<string> {
+  let start, end = 0
+  let partial = ''
+
+  // @ts-ignore
+  for await (const chunk of stream) {
+    const full = chunk+partial as string
+    for (const char of full) {
+      if (char === '\n') {
+        yield chunk.substring(start, end)
+        end++
+        start = end
+      } else {
+        end++
+      }
+    }
+
+    partial = chunk.substring(start, end)
+    start = 0
+    end = 0
+  }
+
+  if (partial.length > 0) {
+    yield partial
+  }
+}
+
+type Batch = {
+  id: string 	// unique identifier for the batch
+  shard?: string 	// assigned shard
+  ts: number 	// timestamp of the event
+
+  format: Format
+  data: unknown
+}
+
+type JsonStream = Batch & {
+  format: Format.JSON_STREAM,
+  data: ReadableStream<Uint8Array>
+}
+
+enum Format {
+  JSON_STREAM = 'json_stream', //jsonl
+}
+
+export class PipelineTransformImpl extends entrypoints.WorkerEntrypoint {
+  #batch?: Batch
+  #initalized: boolean = false
+
+  constructor(ctx: unknown, env: unknown) {
+    super(ctx, env)
+  }
+
+  // stub overriden on the sub class
+  // @ts-ignore
+  public async transformJson(data: object[]): Promise<object[]> {
+    throw new Error('should be implemented by parent')
+  }
+
+  // called by the dispatcher which then calls the subclass methods
+  // @ts-ignore
+  private async _transform(batch: Batch): Promise<JsonStream> {
+    if (this.#initalized) {
+      throw new Error('pipeline entrypoint has already been initialized')
+    }
+
+    // test call to make sure the pipeline entrypoint is setup correctly
+    if (batch === null) {
+      // making sure the function was overriden by an implementing subclass
+      if (this.transformJson !== PipelineTransformImpl.prototype.transformJson) {
+        return null as unknown as JsonStream
+      } else {
+        throw new Error('the transformJson method must be overridden by the PipelineTransform subclass')
+      }
+    } else {
+      this.#batch = batch
+      this.#initalized = true
+
+      switch (this.#batch!.format) {
+        case Format.JSON_STREAM:
+          const data = await this.readJsonStream()
+          const transformed = await this.transformJson(data)
+          return this.sendJson(transformed)
+        default:
+          throw new Error('unsupported batch format')
+      }
+    }
+  }
+
+  private async readJsonStream(): Promise<object[]> {
```

Comment:
Why don't we use private properties supported by javascript language?
```suggestion
  async #readJsonStream(): Promise<object[]> {
```